### PR TITLE
Add support for keeping a backup of storage for last 100 boots

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

File: libraries/AP_Logger/AP_Logger_File.cpp [hunk 1]
```diff
@@ -100,6 +100,10 @@ void AP_Logger_File::Init()
 
     hal.console->printf("AP_Logger_File: buffer size=%u\n", (unsigned)bufsize);
 
+    // We got this far with logger, this is perfect spot
```
Comment (peterbarker) [inline libraries/AP_Logger/AP_Logger_File.cpp:103]: This should probably go into `AP_Vehicle::setup()`.  AFAICS it doesn't have anything to do with dataflash logging.

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 2]
```diff
@@ -92,6 +95,30 @@ bool AP_RAMTRON::init(void)
         return false;
     }
 
+    if (has_backup()) {
+        ramtron_logdata = new uint8_t[get_size()];
+        ramtron_logdata_len = get_size();
+
+        if (ramtron_logdata != nullptr) {
```
Comment (peterbarker) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:102]: We're still going to register an IO process even though it looks like everything has already gone to hell?

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 3]
```diff
@@ -92,6 +95,30 @@ bool AP_RAMTRON::init(void)
...
+            //enter a random fill value, to tell us if 
+            //we failed read bytes from fram
+            memset(ramtron_logdata, 0xAB, get_size());
+            uint16_t log_offset = 0;
+            while (true) {
+                //read 256 bytes at a time
+                if (!_read_data(log_offset, &ramtron_logdata[log_offset], 256)) {
```
Comment (peterbarker) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:109]: What is guaranteeing you have space in `ramtron_logdata` here?

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 4]
```diff
@@ -92,6 +95,30 @@ bool AP_RAMTRON::init(void)
...
+                    break;
+                } else {
+                    log_offset+=256;
+                }
+                if (log_offset >= get_size()) {
+                    break;
+                }
+            }
+        }
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_RAMTRON::data_integrity_tracker, void));
```
Comment (peterbarker) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:119]: We probably ought not be doing this if we fail to get our integrity data successfully.

Comment (bugobliterator) [reply libraries/AP_RAMTRON/AP_RAMTRON.cpp:119]: This integrity tracker is independent to the data we collected above, with integrity tracker we  only record data to the second half of the storage and calculate crc.

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_RAMTRON/AP_RAMTRON.cpp:102]: This integrity tracker is independent to the data we collected above, with integrity tracker we only record data to the second half of the storage and calculate crc. It is unrelated to fram logging.

File: libraries/AP_Logger/AP_Logger_File.cpp (hunk 1 above)
Comment (bugobliterator) [reply libraries/AP_Logger/AP_Logger_File.cpp:103]: ok, I will do that, this location just looked welcoming...

File: libraries/AP_HAL_ChibiOS/hwdef/CubeOrange/hwdef.dat [hunk 5]
```diff
@@ -26,7 +26,11 @@ FLASH_RESERVE_START_KB 128
 # board voltage
 STM32_VDD 330U
 
-define HAL_STORAGE_SIZE 16384
+# the actual size of FRAM is 32K we use approx half 
+# of it, the rest we use to store backup, and checksum
+define HAL_STORAGE_SIZE 16256
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/CubeOrange/hwdef.dat:31]: why take 128 bytes? it seems like more than we need

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 6]
```diff
@@ -217,3 +332,93 @@ bool AP_RAMTRON::write(uint32_t offset, const uint8_t *buf, uint32_t size)
...
+        return;
+    }
+
+    // create the log directory if need be
+    int ret;
+    struct stat st;
+    const char* _fram_log_directory = "/APM/FRAM";
+
+    EXPECT_DELAY_MS(3000);
+    ret = AP::FS().stat(_fram_log_directory, &st);
```
Comment (tridge) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:366]: better to just unconditionally call the mkdir()

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 7]
```diff
@@ -217,3 +332,93 @@ bool AP_RAMTRON::write(uint32_t offset, const uint8_t *buf, uint32_t size)
...
+
+    fd = AP::FS().open(fname, O_WRONLY|O_CREAT|O_TRUNC);
+    if (fd != -1) {
+        char buf[10];
+        snprintf(buf, sizeof(buf), "%u\r\n", (unsigned)curr_log);
+        const ssize_t to_write = strlen(buf);
+        const ssize_t written = AP::FS().write(fd, buf, to_write);
+        AP::FS().close(fd);
+        if (written < to_write) {
+            return;
```
Comment (tridge) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:401]: leaking fname

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 8]
```diff
@@ -217,3 +332,93 @@ bool AP_RAMTRON::write(uint32_t offset, const uint8_t *buf, uint32_t size)
...
+        }
+    } else {
+        return;
```
Comment (tridge) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:404]: leaking fname

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 9]
```diff
@@ -217,3 +332,93 @@ bool AP_RAMTRON::write(uint32_t offset, const uint8_t *buf, uint32_t size)
...
+    }
+
+    // create and write fram data to file
+    free(fname);
+    ret = asprintf(&fname, "%s/fram_%d.bin", _fram_log_directory, curr_log);
```
Comment (tridge) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:409]: assign fname to nullptr before call

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 10]
```diff
@@ -217,3 +332,93 @@ bool AP_RAMTRON::write(uint32_t offset, const uint8_t *buf, uint32_t size)
...
+    }
+    fd = AP::FS().open(fname, O_WRONLY|O_CREAT|O_TRUNC);
+    free(fname);
+    if (fd != -1) {
+        //finally dump the fram data
+        AP::FS().write(fd, _shadow_buf, HAL_STORAGE_SIZE);
+        AP::FS().write(fd, _backup_buf, HAL_STORAGE_SIZE);
+        AP::FS().close(fd);
+    }
+#else
```
Comment (tridge) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:421]: why the #else return ?

File: libraries/AP_RAMTRON/AP_RAMTRON.cpp [hunk 11]
```diff
@@ -91,10 +97,103 @@ bool AP_RAMTRON::init(void)
...
+                    // calculate and update cksum
+                    cksum = cksum_cksum16(CKSUM_INITIAL_MARKER, _shadow_buf, _storage_size);
+                    write_cksum(cksum);
+                } else {
+                    // we have a healthy backup, we can use it for restore
+                    _healthy_backup = true;
+                }
+                _cksum = cksum;
+            } else {
+                delete _shadow_buf;
```
Comment (tridge) [inline libraries/AP_RAMTRON/AP_RAMTRON.cpp:140]: delete calls should be array delete, like this: delete [] _shadow_buf;

**Comment by bugobliterator** [comment]
> a really key part of this PR is how to test it. How will we induce corruption so we know that the restore works?
The way I tested this was by changing to older firmware release with different Vehicle type (with minor modification to flash size), and see if the restore works when returning back to newer firmware.

File: libraries/AP_HAL_ChibiOS/Storage.cpp [hunk 12]
```diff
@@ -126,12 +126,69 @@ void Storage::_save_backup(void)
...
+    // create the backup directory if need be
+    int ret;
+    const char* _storage_bak_directory = HAL_STORAGE_BACKUP_FOLDER;
+
+    EXPECT_DELAY_MS(3000);
+
+    // We want to do this desperately,
+    // So we keep trying this for a second
+    uint32_t start_millis = AP_HAL::millis();
+    while(!sdcard_retry() && (AP_HAL::millis() - start_millis) < 1000) {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Storage.cpp:138]: don't do any delays if in watchdog reset, using hal.util->was_watchdog_armed()

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/Storage.cpp [hunk 13]
```diff
@@ -126,12 +126,69 @@ void Storage::_save_backup(void)
...
+    if (fname == nullptr && (ret <= 0)) {
+        return;
+    }
+    int fd = AP::FS().open(fname, O_RDONLY);
+    if (fd != -1) {
+        char buf[10];
+        memset(buf, 0, sizeof(buf));
+        if (AP::FS().read(fd, buf, sizeof(buf)-1) > 0) {
+            //only record last 100 backups
+            curr_bak = (strtol(buf, NULL, 10) + 1)%100;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Storage.cpp:159]: instead of hard-coding 100, add a HAL_STORAGE_BACKUP_COUNT, with #ifndef/#define

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
thanks for this Sid!

**Pull request merged**